### PR TITLE
CLN: Unify number recognition tests for all parsers

### LINKED ISSUE #38926: CLN: Unify number recognition tests in read_csv for all parsers
Follow up for #38420 
We should parametrize over all parsers for the test sets to ensure a consistent behavior

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #38926
- [x] tests added / passed
- [x] Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them

Minor regex improvements.
Is a fixture the right thing to do here?

### CODE REVIEW

File: pandas/io/parsers.py [hunk 1]
```diff
@@ -2349,12 +2349,12 @@ def __init__(self, f: Union[FilePathOrBuffer, List], **kwds):
 
         decimal = re.escape(self.decimal)
         if self.thousands is None:
-            regex = fr"^\-?[0-9]*({decimal}[0-9]*)?([0-9](E|e)\-?[0-9]*)?$"
+            regex = fr"^[\-|\+]?[0-9]*({decimal}[0-9]*)?([0-9]?(E|e)\-?[0-9]+)?$"
```
Comment (jbrockmendel) [inline pandas/io/parsers.py:2352]: is the `|` necessary inside `[]`?

File: pandas/tests/io/parser/conftest.py [hunk 2]
```diff
@@ -148,3 +148,57 @@ def encoding_fmt(request):
...
+        ("-,1e1", -1.0),
+        ("-1,e0", -1.0),
+        ("0,1", 0.1),
+        ("1,", 1.0),
+        (",1", 0.1),
+        ("-,1", -0.1),
+        ("1_,", 1.0),
+        ("1_234,56", 1234.56),
+        ("1_234,56e0", 1234.56),
+        # negative cases; must not parse as float
```
Comment (jbrockmendel) [inline pandas/tests/io/parser/conftest.py:172]: can you add something in the docstring about this, i.e. how to interpret the tuple

File: pandas/io/parsers.py (hunk 1 above)
Comment (phofl) [reply pandas/io/parsers.py]: No, thanks.

File: pandas/tests/io/parser/conftest.py (hunk 2 above)
Comment (phofl) [reply pandas/tests/io/parser/conftest.py:172]: Good point, done

**Comment by jreback** [comment]
very nice @phofl

**Pull request merged**